36 - Single product template

mehigh opened the pull request:
This is left: 
- Top carousel
- mobile fixed CTA overlay
- mobile content product details / price size / inlining

And of course copying the product.css from ampstart (i've setup that repo next to this and referenced the CSS with a regular CSS link during development).
Reached 44KB of CSS until now.

![single-product](https://user-images.githubusercontent.com/367593/38781088-fb6e8e9e-40e8-11e8-8ad7-f1aa447420e9.jpg)

mehigh committed:
36 - Single product HTML

miina committed:
Start adding theme implementation for single-adventure.php

miina committed:
Fix phpcs.

miina committed:
Fix CS.

miina commented:
**Started adding theme integration**
Currently (within 9d60672) using placeholders for `header.php`, `footer.php`, and static `comments.php` from @mehigh's created HTML template.

Maybe `comments.php` can be implemented within this PR since that's directly related to the CPT template but perhaps proper header template(s) and footer template could be separate PR-s/issues? Thoughts, @kienstra and @postphotos?

miina committed:
Add method comment.

kienstra commented:
**Header And Footer Templates**

Hi @miina,
>perhaps proper header template(s) and footer template could be separate PR-s/issues

That's a good idea. Separate issues for `header.php` and `footer.php` would help organize this. 

And we could merge the templates into `develop` as soon as they're ready, instead of waiting for all of them.

miina committed:
Merge develop & resolve conflicts.

DavidCramer commented:
@miina - please take a look at #38 as it could affect the structures.

miina committed:
Add functionality for comments. Refactor templates.

miina committed:
Add nonce based on default comments saving. Show rating empty by default.

miina committed:
Add ampState for price.

miina commented:
@kienstra Forgot to mention that this PR also fixes #8 (Ratings logic).

kienstra commented on templates/footer-search.php:
```diff
@@ -0,0 +1,24 @@
+
+?>
+<!-- Search -->
+<section class='travel-search py4 xs-hide sm-hide relative'>
```
This isn't a blocker, but it'd be nice to have double quotes in the template. We can go back and address this after the demo on Thursday.

kienstra commented on templates/entry/meta/reviews-count.php:
```diff
@@ -0,0 +1,11 @@
+$reviews = wp_count_comments( get_the_ID() );
+?>
+<?php /* translators: %d: Number of reviews. */ ?>
+<p class="travel-results-result-subtext"><?php echo sprintf( esc_html__( '%d Reviews', 'travel' ), esc_html( $reviews->approved ) ); ?></p>
```
This could be `printf()` instead of `echo sprintf()`.

kienstra commented on templates/entry/meta/rating.php:
```diff
@@ -0,0 +1,16 @@
+?>
+<div class="relative h2 line-height-2 mb1">
+	<div class="travel-results-result-stars green">
+		<?php for ( $i = 0; $i < $rating; $i++ ) : ?>
```
How about something like:
```php
echo esc_html( str_repeat( '*', $rating ) );
```

kienstra commented on templates/footer-search.php:
```diff
@@ -0,0 +1,24 @@
+<!-- Search -->
+<section class='travel-search py4 xs-hide sm-hide relative'>
+	<div class='px1 md-px2 pb1 relative'>
+		<h3 class='travel-search-heading travel-spacing-none h1 bold mb2 center'><?php esc_html_e( 'Have a specific destination in mind?' ); ?></h3>
```
Could this include the text domain of `'travel'`, and likewise for the other translation functions in the file?
```php
esc_html_e( 'Have a specific destination in mind?', 'travel' )
```

kienstra commented on templates/entry/full.php:
```diff
@@ -0,0 +1,50 @@
+			<option value="1"><?php esc_html_e( 'For 1' ); ?></option>
+			<?php for ( $i = 2; $i <= 8; $i += 2 ) : ?>
+				<?php /* translators: %d: The number of people */ ?>
+				<option value="<?php echo esc_attr( $i ); ?>"><?php echo sprintf( esc_html__( 'For %d', 'travel' ), esc_html( $i ) ); ?></option>
```
This is a minor point, but this could be `printf` instead of `echo sprintf`.

kienstra commented on templates/entry/full.php:
```diff
@@ -0,0 +1,50 @@
+			<?php endfor; ?>
+		</select>
+
+		<a href="#" class="product-cta-btn ampstart-btn rounded center bold white block col-12"><?php esc_html_e( 'Call to book' ); ?></a>
```
Could this please use the text domain of `'travel'`?

kienstra commented on templates/similar-adventures.php:
```diff
@@ -0,0 +1,11 @@
+
+?>
+<!-- Popular -->
+<?php echo amp_travel_render_similar_adventures(); // WPCS: XSS ok. ?>
```
Thanks for creating this template tag.

kienstra commented on footer.php:
```diff
@@ -0,0 +1,81 @@
+ * @package WPAMPTheme
+ */
+
+// Ignore the issues of this file since that's just a copied HTML placeholder.
```
Though this `footer.php` is only a placeholder, it looks good. We could work on dynamic data later, at least after the screencast this Thursday.

kienstra commented on includes/functions.php:
```diff
@@ -13,3 +15,264 @@
+			<span class="travel-results-result-subtext mr1">' .
+			/* translators: %d: The number of reviews */
+			sprintf( esc_html__( '%d Reviews', 'travel' ), esc_html( $comments->approved ) ) . '</span>
+				<span class="travel-results-result-subtext"><svg class="travel-icon" viewBox="0 0 77 100"><g fill="none" fillRule="evenodd"><path stroke="currentColor" strokeWidth="7.5" d="M38.794 93.248C58.264 67.825 68 49.692 68 38.848 68 22.365 54.57 9 38 9S8 22.364 8 38.85c0 10.842 9.735 28.975 29.206 54.398a1 1 0 0 0 1.588 0z"></path><circle cx="38" cy="39" r="10" fill="currentColor"></circle></g></svg>
```
The [plugin's validator](https://github.com/Automattic/amp-wp/blob/develop/includes/utils/class-amp-validation-utils.php) raised `invalid_attribute` errors for`fillRule` and `strokeWidth`.

Similar errors appear on copying this entire `<svg>` into the [AMP validator](https://validator.ampproject.org/):
>The attribute 'fillrule' may not appear in tag 'g'.
>The attribute 'strokewidth' may not appear in tag 'path'

The page is still valid AMP, but the preprocessor stripped those attributes.

kienstra commented on templates/entry/meta/location.php:
```diff
@@ -0,0 +1,16 @@
+$location  = ! empty( $locations ) ? $locations[0] : '--';
+?>
+<p class="travel-results-result-subtext mb1">
+	<svg class="travel-icon" viewBox="0 0 77 100"><g fill="none" fillrule="evenodd"><path stroke="currentColor" strokewidth="7.5" d="M38.794 93.248C58.264 67.825 68 49.692 68 38.848 68 22.365 54.57 9 38 9S8 22.364 8 38.85c0 10.842 9.735 28.975 29.206 54.398a1 1 0 0 0 1.588 0z"></path><circle cx="38" cy="39" r="10" fill="currentColor"></circle></g></svg>
```
As [mentioned](https://github.com/xwp/travel/pull/39/files#r182176668), it looks like `fillrule` and `strokewidth` aren't allowed. The plugin's preprocessor removes them, so the page is still valid AMP. But just so we're aware that they're not appearing on the front-end.

kienstra commented on comments.php:
```diff
@@ -0,0 +1,56 @@
+if ( post_password_required() ) {
+	return;
+}
+
```
This is outside this PR's diff, but [could you please add](https://github.com/xwp/travel/blob/111aa870a7c1f09765a683582de1eefdfe4e08b6/includes/class-amp-travel-cpt.php#L158) `'comments'` to the `'adventure'` CPT `'supports'` argument? I think this is blocking the "Popular Adventures" block from displaying, as you can't add ratings (comments) to the 'adventure' post type.

kienstra commented on includes/functions.php:
```diff
@@ -13,3 +15,264 @@
+			<option value="">--</option>';
+
+	for ( $i = 5; $i >= 1; $i-- ) {
+		echo '<option value="' . esc_attr( $i ) . '">' . sprintf( '%d star(s)', esc_attr( $i ) ) . '</option>';
```
This is looking good. What do you think of this small change to separate the singular and plural?

Instead of 
```php
'%d star(s)'
```
how about:
```php
_n( '%d star', '%d stars', $i, 'travel' )
```

kienstra commented on templates/entry/full.php:
```diff
@@ -0,0 +1,50 @@
+
+	<div class="product-cta lg-col-2 lg-right-align">
+		<h4 class="product-cta-title mb2 line-height-1" [text]="'$' + adventure_current_price">$<?php echo esc_html( get_post_meta( get_the_ID(), 'amp_travel_price', true ) ); ?></h4>
+		<select class="select-arr mb1 rounded" name="count" on="change:AMP.setState({ adventure_current_price: event.value * adventure_price })">
```
Great job with updating the price based on the selection.

<img width="1432" alt="price-updates" src="https://user-images.githubusercontent.com/4063887/38892391-b51c7e78-424c-11e8-8f85-77ecf37f4b28.png">

kienstra committed:
Merge in develop, delete editor-blocks.js.

The develop branch deletes that file,
so having changes in that file in this PR
caused merge conflics.
Please do npm run build when checking this out.

kienstra commented:
**Building editor-blocks.js**

The `develop` branch removes `editor-blocks.js`, in order to avoid merge conflicts in PRs. So I also removed that file from this branch. 

Please do `npm run build` when checking out this branch to use the Gutenberg blocks.

mehigh committed:
36 - top hero carousel

mehigh committed:
36 - Product details adjustments for responsiveness

mehigh committed:
36 - Sticky call to action

mehigh committed:
36 - updated CSS for single product

mehigh committed:
36 - copied images from ampstart travel template

mehigh commented:
The html/css is complete for the page now:
![laptop-view](https://user-images.githubusercontent.com/367593/38920845-ee666ea4-42fc-11e8-9d14-1c5fa49d7835.jpg)
![phone-view](https://user-images.githubusercontent.com/367593/38920848-f0e49886-42fc-11e8-8ff1-fb2130a6307b.jpg)

mehigh committed:
36 - fixed validation error

miina committed:
Template updates.

miina committed:
Merge branch 'feature/36-adventure-single' of github.com:xwp/travel into feature/36-adventure-single

miina committed:
Add sticky CTA.

miina committed:
Fixes from review.

miina committed:
Fix popular posts.

miina committed:
Fix popular posts.

miina commented on includes/functions.php:
```diff
@@ -13,3 +15,268 @@
+
+	for ( $i = 5; $i >= 1; $i-- ) {
+		/* translators: %d: Rating */
+		echo '<option value="' . esc_attr( $i ) . '">' . sprintf( _n( '%d star', '%d stars', $i, 'travel' ), number_format_i18n( $i ) ) . '</option>';
```
@kienstra This line is now throwing CS error:
`includes/functions.php:153:67: error - All output should be run through an escaping function (see the Security sections in the WordPress Developer Handbooks), found '_n'. (WordPress.XSS.EscapeOutput.OutputNotEscaped)`.

Any thoughts how to fix that?

kienstra replied:
Hi @miina, wrapping the `_n()` call in `esc_html()` should take care of that. Otherwise:
```php
// WPCS: XSS ok.
```

miina commented:
@kienstra Thanks for reviewing, most of the issues should be addressed now. Left the gallery as static for now, hopefully that's OK for now and could be addressed later if needed. There's just [one issue](https://github.com/xwp/travel/pull/39/files#r182389789) that I'm not sure how to fix, any thoughts?

@mehigh Thanks for generating the HTML!
There's one CSS issue that would be great if you could help with -- namely within the Similar Posts section there's a div with class `travel-overflow-container`. The CSS for the class in the `adventure.css` looks like this:
```
.travel-overflow-container {
    display: inline-block;
    min-width: 100vw;
    padding-left: calc((100vw - 72rem) / 2);
    padding-right: calc((100vw - 72rem) / 2);
}
```

However, with the AMP plugin on it looks like the padding gets stripped and only this is left:
```
.travel-overflow-container {
    display: inline-block;
    min-width: 100vw;
}
```
Would be great if you could help with that issue.

miina committed:
Fix phpcs.

mehigh committed:
36 - Added a class to work around the calc

mehigh committed:
36 - added workaround to failing calc styles

mehigh committed:
Merge branch 'feature/36-adventure-single' of github.com:xwp/travel into feature/36-adventure-single

kienstra committed:
Add esc_html to address Travis issue.

It may be that number_format_i18n() needs esc_html().
WordPress.XSS.EscapeOutput.OutputNotEscaped

miina committed:
Style fixes.

miina committed:
Merge branch 'feature/36-adventure-single' of github.com:xwp/travel into feature/36-adventure-single

kienstra reviewed (approved):
**Approved**

Hi @miina and @mehigh,
Thanks for your great work here. The single 'adventure' template looks good.

postphotos commented:
Making sure we capture - related to #36.
